sql/src/test/java/org/apache/druid/sql/calcite/CalciteArraysQueryTest.java

```diff
@@ -4807,4 +4808,187 @@ public void testUnnestWithGroupByHavingWithWhereOnUnnestCol()
+                      join(
+                          new QueryDataSource(
+                              Druids.newScanQueryBuilder()
+                                    .dataSource(
+                                        UnnestDataSource.create(
+                                            new TableDataSource(CalciteTests.DATASOURCE1),
+                                            expressionVirtualColumn("j0.unnest", "\"dim3\"", ColumnType.STRING),
+                                            null
+                                        )
+                                    )
+                                    .intervals(querySegmentSpec(Filtration.eternity()))
+                                    .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
+                                    .legacy(false)
+                                    .columns(
+                                        "__time",
+                                        "cnt",
+                                        "dim1",
+                                        "dim2",
+                                        "dim3",
+                                        "j0.unnest",
+                                        "m1",
+                                        "m2",
+                                        "unique_dim1"
+                                    )
+                                    .context(QUERY_CONTEXT_UNNEST)
+                                    .build()
+                          ),
+                          new QueryDataSource(
+                              Druids.newScanQueryBuilder()
+                                    .dataSource(new TableDataSource(CalciteTests.DATASOURCE1))
+                                    .intervals(querySegmentSpec(Filtration.eternity()))
+                                    .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
+                                    .legacy(false)
+                                    .context(QUERY_CONTEXT_UNNEST)
+                                    .columns("__time", "cnt", "dim1", "dim2", "dim3", "m1", "m2", "unique_dim1")
+                                    .build()
+                          ),
+                          "_j0.",
+                          "1",
+                          JoinType.INNER,
+                          null
+                      )
+                  )
+                  .intervals(querySegmentSpec(Filtration.eternity()))
+                  .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
+                  .legacy(false)
+                  .columns("j0.unnest")
+                  .context(QUERY_CONTEXT_UNNEST)
+                  .build()
+        ),
+        NullHandling.sqlCompatible() ?
+        ImmutableList.of(
+            new Object[]{"a"},
+            new Object[]{"a"},
+            new Object[]{"a"},
+            new Object[]{"a"},
+            new Object[]{"a"},
+            new Object[]{"a"},
```

Comment:
Refactoring comment - We should use a loop or a method that supports adding multiple elements at a single time.

I think one of the concise approach can be something like:
```java
ImmutableList.Builder<String> resultsBuilder = ImmutableList.builder();
resultsBuilder.addAll(Collections.nCopies(6, new Object[]{"a"}));
resultsBuilder.addAll(Collections.nCopies(12, new Object[]{"b"}));
....
resultsBuilder.addAll(Collections.nCopies(12, new Object[]{NullHandling.defaultStringValue()}));


```